localstack/services/sqs/provider.py

```diff
@@ -0,0 +1,1243 @@
+        result: Dict[MessageSystemAttributeName, str] = {
+            MessageSystemAttributeName.SenderId: context.account_id,
+            MessageSystemAttributeName.SentTimestamp: str(now()),
+        }
+
+        if message_system_attributes is not None:
+            for attr in message_system_attributes:
+                result[attr] = message_system_attributes[attr]["StringValue"]
+
+        return result
+
+    def _validate_queue_attributes(self, attributes: QueueAttributeMap):
+        valid = [k[1] for k in inspect.getmembers(QueueAttributeName)]
+
+        for k in attributes.keys():
+            if k not in valid:
+                raise InvalidAttributeName("Unknown Attribute %s" % k)
+
+    def _validate_actions(self, actions: ActionNameList):
+        service = load_service(service=self.service, version=self.version)
+        # FIXME: this is a bit of a heuristic as it will also include actions like "ListQueues" which is not
+        #  associated with an action on a queue
+        valid = list(service.operation_names)
+        valid.append("*")
+
+        for action in actions:
+            if action not in valid:
+                raise InvalidParameterValue(
+                    f"Value SQS:{action} for parameter ActionName is invalid. Reason: Please refer to the appropriate "
+                    "WSDL for a list of valid actions. "
+                )
+
+    def _assert_permission(self, context: RequestContext, queue: SqsQueue):
+        action = context.operation.name
+        account_id = context.account_id
+
+        if account_id == queue.owner:
+            return
+
+        for permission in queue.permissions:
+            if permission.account_id != account_id:
+                continue
+            if permission.action == "*":
+                return
+            if permission.action == action:
+                return
+
+        raise CommonServiceException("AccessDeniedException", "Not allowed (TODO: correct message)")
+
+    def _assert_batch(self, batch: List):
+        if not batch:
+            raise EmptyBatchRequest
+        visited = set()
+        for entry in batch:
+            # TODO: InvalidBatchEntryId
+            if entry["Id"] in visited:
+                raise BatchEntryIdsNotDistinct()
+            else:
+                visited.add(entry["Id"])
+
+    def _require_queue_by_arn(self, dead_letter_target_arn):
+        arn_parts = dead_letter_target_arn.split(":")
+        url = "{}/{}/{}".format(
+            get_edge_url(), arn_parts[len(arn_parts) - 2], arn_parts[len(arn_parts) - 1]
+        )
+        queue = self._require_queue_by_url(url)
+        return queue
+
+
+def _create_mock_sequence_number():
+    return "".join(random.choice(string.digits) for _ in range(20))
+
+
+# Method from moto's attribute_md5 of moto/sqs/models.py, separated from the Message Object
+def _create_message_attribute_hash(message_attributes):
+
+    if not isinstance(message_attributes, dict):
+        return
```

Comment:
I would need to check again but if I recall correctly this was able to be null or some empty list. The None return value was important however because if it is indeed not a dict we do not want return anything, omitting it completely from the response (we wouldn't want to return any form of hash for attributes that were never passed to begin with).
However it would probably better to not call the function at all in that case. Guess it's a matter of where that check is less intrusive

Reply:
I put an explanation there for now. Having to check the input before every invocation seems tedious as well

ToDiscuss

Reply:
added an optional return type and comment to make things clearer